radix-engine/src/model/access_controller/state_machine.rs

```diff
@@ -0,0 +1,421 @@
+        match self.state {
+            (_, ref mut primary_operations_state @ PrimaryOperationState::Normal, _)
+                if input.proposer == Proposer::Primary =>
+            {
+                // Transition the primary operations state from normal to recovery
+                *primary_operations_state = PrimaryOperationState::Recovery(input.proposal);
+                Ok(())
+            }
+            (_, _, ref mut recovery_operations_state @ RecoveryOperationState::Normal)
+                if input.proposer == Proposer::Recovery =>
+            {
+                match self.timed_recovery_delay_in_minutes {
+                    Some(delay_in_minutes) => {
+                        let current_time = Runtime::sys_current_time(api, TimePrecision::Minute)?;
+                        let timed_recovery_allowed_after = current_time
+                            .add_minutes(delay_in_minutes as i64)
+                            .map_or(access_controller_runtime_error!(TimeOverflow), |instant| {
+                                Ok(instant)
+                            })?;
+
+                        *recovery_operations_state =
+                            RecoveryOperationState::Recovery(RecoveryRecoveryState::Timed {
+                                proposal: input.proposal,
+                                timed_recovery_allowed_after,
+                            });
+                        Ok(())
+                    }
+                    None => {
+                        *recovery_operations_state = RecoveryOperationState::Recovery(
+                            RecoveryRecoveryState::Untimed(input.proposal),
+                        );
+                        Ok(())
+                    }
+                }
+            }
+            _ => Err(RuntimeError::ApplicationError(
+                ApplicationError::AccessControllerError(
+                    AccessControllerError::RecoveryForThisProposerAlreadyExists {
+                        proposer: input.proposer,
+                    },
+                ),
+            )),
+        }
+    }
+}
+
+pub(super) struct AccessControllerQuickConfirmRecoveryStateMachineInput {
+    pub confirmor: Role,
+    pub proposer: Proposer,
+    pub proposal_to_confirm: RecoveryProposal,
+}
+
+impl TransitionMut<AccessControllerQuickConfirmRecoveryStateMachineInput>
+    for AccessControllerSubstate
+{
+    type Output = RecoveryProposal;
+
+    fn transition_mut<Y>(
+        &mut self,
+        _api: &mut Y,
+        input: AccessControllerQuickConfirmRecoveryStateMachineInput,
+    ) -> Result<Self::Output, RuntimeError>
+    where
+        Y: SystemApi + EngineApi<RuntimeError> + InvokableModel<RuntimeError>,
+    {
+        // This transition can not be performed when the confirmor and the proposer are the same.
+        // As in, Role X can't perform a quick confirm on a recovery initiate by Role X, it has to
+        // be different.
+        if input.confirmor == input.proposer.into() {
```

Comment:
Rather than this being implemented as logic, this part could also be cleaned up if the quick_confirm transitions were implemented separately.

Reply:
This would also allow us to remove the need for the `QuickConfirmAs[Role]` pattern we currently have.

Reply:
Could you expand on this a bit? I think that somewhere in the logic we would need to check to ensure that role X isn't confirming recovery initiated by role X. Not sure how implementing the transitions separately achieves that.

Reply:
Yeah, we might want to separate this out into multiple state machine inputs (and probably also do the same for other inputs as well). Only concern here is that it seems like we've formed an hour glass. Multiple invocations go to the same executor which then map out to different state machine inputs. 

```
╲      ╱     Different Invocations
 ╲    ╱
  ╲  ╱
  │  │       Same Executor
  ╱  ╲
 ╱    ╲
╱      ╲     Different State Machine Inputs
```

We could switch to one executor and one input per invocation to clean up the logic a little bit but having a single executor is nice for stuff like updating the access rules and having less duplicated logic.

Reply:
I think one input per invocation makes more sense. Less duplicated logic can be solved by using functions/methods encapsulating the duplicated logic.

Reply:
Sounds good, I've refactored the code to use one executor per invocation and one input per executor (i.e. no more hourglass)